core/common/byte_stream.hpp

```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_CORE_COMMON_BYTE_STREAM_HPP_
+#define KAGOME_CORE_COMMON_BYTE_STREAM_HPP_
+
+#include <optional>
+#include <outcome/outcome.hpp>
+
+namespace kagome::common {
+  /**
+   * @brief interface for Decoders
+   */
+  class ByteStream {
+   public:
+    enum class AdvanceErrc { OUT_OF_BOUNDARIES = 1 };
+
+    virtual ~ByteStream() = default;
+
+    /**
+     * @brief hasMore Checks whether n more bytes are available
+     * @param n Number of bytes to check
+     * @return True if n more bytes are available and false otherwise
+     */
+    virtual bool hasMore(uint64_t n) const = 0;
+
+    /**
+     * @brief nextByte Takes current byte and moves pointer to the next one
+     * @return Current byte
+     */
+    virtual std::optional<uint8_t> nextByte() = 0;
+
+    /**
+     * @brief Shifts pointer by the given distance
+     * @return void if advance successful, error if advance is not possible
+     */
+    virtual outcome::result<void> advance(uint64_t dist) = 0;
+  };
+}  // namespace kagome::common
+
+OUTCOME_HPP_DECLARE_ERROR(kagome::common, ByteStream::AdvanceErrc)
```

Comment:
maybe it makes sense to declare separate error.hpp|.cpp for errors

Reply:
This is the usual error for every class who implement ByteStream interface. If we move it outside the class, then users who will implement byte stream should include both byte_stream.hpp and byte_stream_error.hpp. I would like to avoid that if possible

Reply:
Although if you insist on that, we can move error to separate file

Reply:
I don't insist, In this scope maybe it's ok to have error defined in byte_stream.hpp  and byte_stream.cpp. I don't mind if you feel it's ok. 
When there are many files each having own error definition, in this case having all definitions combined in a certain error.cpp file looks desirable. Combining all errors in shared error.hpp looks not necessary, but still can be useful.